File: bindings/java-jni/src/main/java/org/maplibre/nativejni/geo/FeatureIdentifier.java
package org.maplibre.nativejni.geo;

import java.util.Objects;

/** GeoJSON feature identifier value. */
public sealed interface FeatureIdentifier
    permits FeatureIdentifier.Null,
        FeatureIdentifier.UInt,
        FeatureIdentifier.Int,
        FeatureIdentifier.DoubleValue,
        FeatureIdentifier.StringValue,
        FeatureIdentifier.Unknown {
  static Null nullValue() {
    return Null.INSTANCE;
  }

  static UInt unsigned(long value) {
    return new UInt(value);
  }

  static Int of(long value) {
    return new Int(value);
  }

  static DoubleValue of(double value) {
    return new DoubleValue(value);
  }

  static StringValue of(String value) {
    return new StringValue(value);
  }

  static Unknown unknown(int rawType) {
    return new Unknown(rawType);
  }

  /** Singleton null identifier. */
  final class Null implements FeatureIdentifier {
    public static final Null INSTANCE = new Null();

    private Null() {}

    @Override
    public boolean equals(Object other) {
      return other instanceof Null;
    }

    @Override
    public int hashCode() {
      return 0;
    }

    @Override
    public String toString() {
      return "FeatureIdentifier.Null";
    }
  }

  /**
   * Unsigned 64-bit feature identifier stored in a Java {@code long} using the native bit pattern.
   *
   * <p>Use {@link Long#compareUnsigned(long, long)} and {@link Long#toUnsignedString(long)} when
   * interpreting this value as unsigned.
   */
  record UInt(long value) implements FeatureIdentifier {}

  record Int(long value) implements FeatureIdentifier {}

  record DoubleValue(double value) implements FeatureIdentifier {}

  record StringValue(String value) implements FeatureIdentifier {
    public StringValue {
      Objects.requireNonNull(value, "value");
    }
  }

  /** Feature identifier with an unknown native type tag returned by a newer C API. */
  record Unknown(int rawType) implements FeatureIdentifier {}
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/geo/FeatureIdentifier.kt
package org.maplibre.nativeffi.geo

/** GeoJSON feature identifier value. */
public sealed interface FeatureIdentifier {
  public data object Null : FeatureIdentifier

  /**
   * Unsigned feature identifier (`uint64_t` in the C ABI) stored as a [Long] bit pattern.
   *
   * Compare and format as unsigned by converting [value] with [Long.toULong].
   */
  public data class UInt(public val value: Long) : FeatureIdentifier

  /**
   * Signed feature identifier (`int64_t` in the C ABI). Name mirrors the C discriminant, not Kotlin
   * [kotlin.Int].
   */
  public data class Int(public val value: Long) : FeatureIdentifier

  public data class DoubleValue(public val value: Double) : FeatureIdentifier

  public data class StringValue(public val value: String) : FeatureIdentifier

  public class Unknown internal constructor(public val rawType: kotlin.Int) : FeatureIdentifier {
    override fun equals(other: Any?): Boolean = other is Unknown && rawType == other.rawType

    override fun hashCode(): kotlin.Int = rawType

    override fun toString(): String = "Unknown(rawType=$rawType)"
  }
}
